### DISCUSSION TITLE
Single source Google Calendar to be split into multiple calendars

### DISCUSSION THREAD

**Discussion by tim9hop** [derekantrican/GAS-ICS-Sync#461] [Q&A] [answered]
I have a single source Google calendar that I want to split out into multiple Google calendars based on certain keywords.

The source calendar is based on a subscription URL to a golf management application which has a calendar of events. 

Each event contains a prefix in the title for the section of the golf club that they relate to ie. Men's, Ladies, Seniors etc. followed by the type of competition. As an example "Seniors' Texas Scramble".

The source calendar is not provided by section so this needs to be split up into multiple calendars so that members can subscribe just to the calendar they are interested in. This could be done using a filter with key words ie. 'Seniors' to create a separate Google calendar.

Has anyone created anything similar to this..?

**Comment by Lonestarjeepin** [comment, marked as answer]
If you switch to the settings-v2 branch, you can accomplish this. Just list your source calendar multiple times, apply different filters to each entry, and different target calendar to each entry, and you're good to go.

**Reply by tim9hop** [reply]
Thanks for the quick response.

**Reply by tim9hop** [reply]
I have a question regarding the filter. I need to filter on the Title of the event or at least the first x characters.
Is this the correct parameter and format..?

'onlyConfirmed': {
                      parameter: "summary",       // Include events whose summary starts with "Seniors".
                      type: "include",
                      comparison: "equals",
                      criterias: ["^Seniors:"]
                    },

**Reply by jonas0b1011001** [reply]
```
'onlyConfirmed': {
parameter: "summary", // Include events whose summary starts with "Seniors".
type: "include",
comparison: "begins with",
criterias: ["Seniors:"]
},
```
or
```
'onlyConfirmed': {
parameter: "summary", // Include events whose summary starts with "Seniors".
type: "include",
comparison: "regex",
criterias: ["^Seniors:"]
},
```

**Reply by tim9hop** [reply]
Thanks for the answer. 

I don't see the 'Debug' area mentioned in the About file. 

Has the screen changed since this was written or am I looking in the wrong place..?

**Reply by jonas0b1011001** [reply]
![image](https://github.com/user-attachments/assets/e852a61d-8119-4611-b2d4-7350a1d88e29)

**Reply by tim9hop** [reply]
Thanks, I've have now created a new project in Apps Script. 
How do I copy or link the files to the project..?

**Reply by jonas0b1011001** [reply]
As this is not an official release yet, you'll have to manually create blank files in the code editor and copy&paste the code from github.

**Comment by tim9hop** [comment]
Thanks again.
I've made some progress but when trying the sync it comes up with errors.

I've included a link to the script below:

(https://script.google.com/d/1EuZgB3kD2GU2Rfq3Rk9W3H_NRJVKRSaLHj6vc50xqZLUf3kJfp5ZcN6r/edit?usp=sharing)

**Reply by jonas0b1011001** [reply]
Can you share the errors?

I don't have permission to view your script, you'd have to change sharing setting to 'anyone'.

**Reply by tim9hop** [reply]
Hopefully this works for you…
https://script.google.com/d/1EuZgB3kD2GU2Rfq3Rk9W3H_NRJVKRSaLHj6vc50xqZLUf3kJfp5ZcN6r/edit?usp=sharing

**Reply by jonas0b1011001** [reply]
It does.

1. You are using the wrong url as source url. Use the 'Secret address in iCal format' instead of the embedded calendar.
2. Remove any filter you don't want to apply.

**Reply by tim9hop** [reply]
ok, I've applied the secret address and it now executes and executes etc..
I added the 28 day limit to try and stop it early but it seems to carry on.
With the filter of 'Seniors' I would expect only about 30 events to be added to the rest of the year.

A couple of questions:
1. Does the sync start from the beginning of the source calendar which is the year 2022 or from today?
2. If I stop the execution, does the whole sync cancel ie. if its halfway through, will I get at least 50% of the events showing as I currently cannot see any on the target calendar.

**Reply by jonas0b1011001** [reply]
1. yes
2. you should see progress. Check the logs at https://script.google.com/home/executions to see where it hangs.

**Reply by tim9hop** [reply]
I've now found the target calendar which its populating which is great...!
However, it's not filtering the events that I requested. It's adding all events....
This is the filter:  

'onlyConfirmed': {
                      parameter: "summary", // Include events whose summary starts with "Seniors'".
                      type: "include",
                      comparison: "begins with",
                      criterias: ["Seniors'"]

I also tried using 'onlyMeetings' instead of 'onlyConfirmed' but it had the same result.
Do you think the other filters are interfering with it so I should try removing them all except this one..?

**Comment by jonas0b1011001** [comment]
First, make sure the filter identifiers are unique, you currently have two filters named `onlyMeetings`, after that it's irrelevant if you have more filters than you need, because you have to actually apply the filter to the source calendar.

Add `filters: ['']` with the identifier of your Seniors filter to the sourceCalendars definition. See the example in settings.gs for how it should look (there currently is a typo though, `filers` should be `filters`)

**Reply by tim9hop** [reply]
Thanks, again have made good progress with 3 calendars produced from the source.
However, the last one (Men) has every event duplicated. Originally, I thought it was because I used ....Test 1 - Men's (with an apostrophe) and thought it might be affecting the duplication if this was a apostrophe was special character so changed it to .....Test 1 - Men.
Could you check if you can see any error causing the duplication please..?

**Reply by jonas0b1011001** [reply]
Script looks fine and does not produce duplicates for me.

Does this happen on each run?
Can you share one of the execution logs from https://script.google.com/home/executions ?

**Comment by tim9hop** [comment]
yes, I get issues on each run..

hopefully you can access this one which I’ve just run…in fact it’s got worse.

https://script.google.com/home/projects/1EuZgB3kD2GU2Rfq3Rk9W3H_NRJVKRSaLHj6vc50xqZLUf3kJfp5ZcN6r/executions

1. Ladies calendar – created twice (plus I notice that it is only picking up 1 event (in August 2024) and ignoring the remainder
2. Men’s calendar – created twice
3. Seniors calendar – created once but each event is duplicated.

Here is the calendar:


<img width="1323" alt="Screenshot 2024-10-25 at 10 18 29" src="https://github.com/user-attachments/assets/078303d2-8022-42c8-8c46-4ed94ecce445">

**Reply by tim9hop** [reply]
In terms of process, I usually delete the new calendars (Men, Ladies etc) from the Google calendar, go to Apps Script and run Uninstall, Install and then the Sync.

**Reply by jonas0b1011001** [reply]
Remove the `//` from this line and let it run again.
https://github.com/derekantrican/GAS-ICS-Sync/blob/839246c3d067a877f3eb998b447e94514bf73f34/Main.gs#L54

In general there is no need to uninstall/install, the triggers will always execute the latest version of your script.

**Reply by tim9hop** [reply]
Thanks, that removes the duplication in both calendars and events.
However, it is still only pulling 1 event for the Ladies despite many events in the future as per this screenshot from the source diary..
<img width="939" alt="Screenshot 2024-10-25 at 10 53 13" src="https://github.com/user-attachments/assets/f906b791-001a-426e-b3c6-f8a26f72e3da">

**Reply by jonas0b1011001** [reply]
These events are all missing from the **source** ics..

Download the ics file, open in a text editor and search for 'Ladies', there's only one match.

**Reply by tim9hop** [reply]
Thanks, you're right.
I have now used the URL from the IG system to provide a new subscription to my own Google account. Then used the public address iCal from this Google calendar in the script. The only problem with this feed is that the events start with today's date and not from the beginning of the year.

**Reply by tim9hop** [reply]
I now need to test that people can subscribe to the created calendars (Men, Ladies etc).

I tried to add a new calendar in my other Google account with the 'subscription' option and paste in the public URL from the created calendar and also shareable link but none of these seem to work.

Can you confirm which option I should take to create the subscription to the created accounts please..?

**Comment by Lonestarjeepin** [comment]
I do something similar for my kids' activities.  I have a calendar for each kid, then I give people (grandparents, etc.) permission in the calendar settings in google in the "Share with specific people or groups" section.  Just specify the recipient email address and they will get notified to subscribe to the calendar.   You could also make it Publicly sharable so you don't have to admin the user list (there's a check box for "Make available to public").  For your comment about entries back to the beginning of the year, I checked the ics for Seniors and it looks like it only contains future events... so the script is behaving correctly.  You might see if your source can enable past events if those are needed.

**Reply by tim9hop** [reply]
Thanks for your help in getting this solution done.
When I get back to the UK in November, I’ll be testing it with the golf club so we can use it for 2025 events.
As a grandparent, I like your idea for sharing calendars with the family.
I’ll see if IG can provide all the events and not just those going forward.

**Comment by jonas0b1011001** [comment]
Don't forget to revert the sharing settings of your script.

**Reply by tim9hop** [reply]
ok, hopefully I've re-set the sharing now..

I also have updated a couple of the files: About.gs and Settings.gs with some comments in yellow (attached in MSWord format) which you can consider. 

As someone unfamiliar with Git, Github and Apps Script, my additional comments may help others and reduce the number of queries....

[GAS-Sync- About.gs.docx](https://github.com/user-attachments/files/17544708/GAS-Sync-.About.gs.docx)
[GAS-Sync-SETTINGS.gs.docx](https://github.com/user-attachments/files/17544709/GAS-Sync-SETTINGS.gs.docx)

**Reply by jonas0b1011001** [reply]
Thanks, appreciate the feedback!

**Reply by tim9hop** [reply]
I notice that there is a 'Deployment' option for the script...
Is it worth using this to run it or just do the install and run the start sync..?

**Reply by jonas0b1011001** [reply]
There's no benefit.
I recommend sticking with what you have.

**Reply by tim9hop** [reply]
thanks, will do.

**Reply by tim9hop** [reply]
Is there a way of using asterisks for my filter criteria instead of words..?
One of my groups would look better with a symbol rather than a name.
I tried putting '**' and also "'**'" and they were both it rejected...

**Reply by jonas0b1011001** [reply]
Try with a backslash `\*`

**Reply by tim9hop** [reply]
I assume you mean: ["\\**"]

tried that, I get: SyntaxError: Invalid regular expression: /^(**)/: Nothing to repeat

I also tried a double quote ""**"" which runs without errors but doesn't find the one event I've set up..

**Reply by tim9hop** [reply]
Actually, I think the source isn't including my event so ignore this for now..

**Reply by jonas0b1011001** [reply]
`criterias: ["\\*"]`

**Reply by tim9hop** [reply]
That works, thanks again..

**Reply by tim9hop** [reply]
Is there a way of having multiple filters where its 'xxxx' or 'yyyy' or 'zzzz' for the same target calendar..?

I tried using a string of filters with a comma separator but it seems to combine the filters ie. uses 'and'...

**Reply by jonas0b1011001** [reply]
Does the scheme from the examples not work?
```
'orExample': {
parameter: "summary", 
type: "include",
comparison: "begins with",
criterias: ["xxxx", "yyyy", "zzzz"]
}
```

**Reply by tim9hop** [reply]
I tried it again and it works ok now, maybe I had the syntax wrong..also tried it with 'regex' and that works well too..thanks.

**Reply by tim9hop** [reply]
I have changed the names of the created calendars and now get an error:

GoogleJsonResponseException: API call to calendar.calendars.insert failed with error: **Calendar usage limits exceeded.**
setupTargetCalendar	@ Helpers.gs:217
startSync	@ Main.gs:73

It may just be a co-incidence and may have exceeded a limit that I don't know about, any suggestions..?

**Reply by jonas0b1011001** [reply]
There are usage limits for the google calendar api. During testing it's possible to exceed these limits.
Your only option is to wait for Google to 'unblock' your account, just give it a few hours and it should resolve itself.

**Comment by tim9hop** [comment]
I am noticing that I get a failed run every other 'time driven' run. Although I've set the time to '15' minutes, it seems to attempt an earlier time then fails. the '15' minute run seems to complete as expected.

![Screenshot 2024-11-24 at 11 32 15](https://github.com/user-attachments/assets/93438103-bdd1-411d-8cb8-90e6901932a7)

I am running this under the 'clandonregisgolfclub@gmail.com' login.

[https://script.google.com/d/1JbJzYMrpY-nFh9WUqoqCfYv8THGtsTIyn_WtoW6VzclI94nYQQYqID_2/edit?usp=sharing](url)

**Reply by jonas0b1011001** [reply]
do you have multiple triggers at https://script.google.com/home/triggers?

**Reply by tim9hop** [reply]
not knowingly, I don't remember creating any....should I just delete them..?
![Screenshot 2024-11-24 at 14 43 28](https://github.com/user-attachments/assets/58a3be09-dd44-4117-9dc7-241dc133f111)

**Reply by jonas0b1011001** [reply]
I'd suggest deleting them all and running install again.

**Reply by tim9hop** [reply]
ok, have deleted them so will let it run for 24hrs and see if they re-occur, thanks.

**Reply by tim9hop** [reply]
Looks like this is working correctly now.

Going fwd, if I change anything in the Settings.gs file, do I need to delete the triggers each time before I install and run the new version..?

**Reply by Lonestarjeepin** [reply]
When you make changes, you really just need to Save (no uninstall or install). Your saved changes will get used on the next scheduled run.
